sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -249,70 +249,107 @@ object DateTimeUtils {
    *         the input string can't be parsed as timestamp, the result timestamp segments are empty.
    */
   def parseTimestampString(s: UTF8String): (Array[Int], Option[ZoneId], Boolean) = {
-    if (s == null) {
+    def isValidDigits(segment: Int, digits: Int): Boolean = {
+      // A Long is able to represent a timestamp within [+-]200 thousand years
+      val maxDigitsYear = 6
+      // For the nanosecond part, more than 6 digits is allowed, but will be truncated.
+      segment == 6 || (segment == 0 && digits > 0 && digits <= maxDigitsYear) ||
+        (segment != 0 && segment != 6 && digits <= 2)
```

Comment:
segments except year are allowed to have 0 digits before this PR. so I didn't do zero checks for other segments.
for example, before this PR, the below query is valid:
```sql
select cast('12::' as timestamp); -- output: 2021-07-07 12:00:00
select cast('T' as timestamp); -- output: 2021-07-07 00:00:00
```

Reply:
good catch, we should fail it. Let's do it in another PR.